wagtail/utils/utils.py

```diff
@@ -40,3 +44,49 @@ def flatten_choices(choices):
             # choice (key, display value)
             ret[str(key)] = value
     return ret
+
+
+def reduce_image_size(avatar, size_bound, max_dimensions=(400, 400)):
+    """
+    Reduce an image's file size to a size_bound in kilobytes.
+
+    :param avatar: The original image gotten from the image field of the form
+    :param size_bound: Desired size in KB
+    :param max_dimensions: Maximum dimensions for resizing
+    """
+    temp_buffer = BytesIO()
+    kilobyte = 1024
+    with Image.open(avatar) as img:
+        if img.mode == "RGBA":
+            img = img.convert("RGB")
+        img.thumbnail(max_dimensions, Image.LANCZOS)
+        img_ext = avatar.name.split(".")[-1]
+
+        quality = 95
+        step = 5
+
+        while quality > 10:
+            temp_buffer.seek(0)
+            img.save(
+                temp_buffer,
+                format="JPEG" if img_ext != "png" else "PNG",
+                quality=quality,
+                optimize=True,
+            )
+            file_size_kb = temp_buffer.tell() / kilobyte
+
+            if file_size_kb <= size_bound:
+                break
+
+            quality -= step
+
+        temp_buffer.seek(0)
+        in_mem_avatar = InMemoryUploadedFile(
+            file=temp_buffer,
+            field_name="ImageField",
+            name=avatar.name,
+            content_type="image/jpeg" if img_ext != "png" else "PNG",
```

Comment:
This should be `"image/png"` rather than `"PNG"`, I think...?

Also, does this correctly handle file types other than jpg/png, such as gif? It looks to me like this will convert the file to JPEG but still save it under the original filename with a .gif extension.

Reply:
Yeah, that edge case wasn't captured, would have to address it and also correct the oversight in the code